### ISSUE TITLE
argparse tab autocompleter

### ISSUE THREAD

**Issue by anselor** [python-cmd2/cmd2#349]
I'm working on a tab completion implementation that automatically generates completions based on an argparse definition. Creating this issue to enable discussions.

**Comment by tleonhardt** [comment]
f we move to the state where we have a 2nd file as part of the cmd2 module instead of just the single cmd2.py file, then we should change the directory structure to create a src directory and make cmd2 a proper Python package by including a init.py and such.

Hopefully this should be a small change, but it will likely require small tweaks in all of the following areas:

packaging for distribution to PyPI
Sphinx documentation - where it looks for the cmd2 source when using autodoc features
unit testing - where it looks for the cmd2 source and what it does to install before running tests

**Comment by tleonhardt** [comment]
When you create a PR for this, make sure you update **setup.py** to add a dependency on ``colorama`` if you use that module. I'm fine with that - it is a very common library which I always end up using anyways.

For a long time I've been considering adding a dependency on ``colorama`` and retrofitting ``cmd2``'s existing colorizing code to use ``colorama`` so that it will work on Windows as well as the POSIX OSes.

**Comment by tleonhardt** [comment]
From the limited experimentation I have done playing with this feature on the **autocompleter** branch, I think it rocks socks.  It is a very slick feature which makes tab completing everything related to ``argparse`` arguments quite simple.  And the way it provides a hint specifically for the argument you are attempting to complete is very helpful.

**Comment by tleonhardt** [comment]
@anselor 
I love this feature.  It combines so synergistically with the existing ``argparse`` decorators and existing tab-completion functionality that it goes a long way to make it just so damn easy to provide awesome tab-completion that end users will love pretty much "out-of-the-box".  

I recommend that you work in conjunction with @kmvanbrunt  to update the "front-page" code example in **README.md** to make use this feature as well as the argument and/or index based tab completion and make that part of the PR that merges this feature in.  Keep it simple, but make it readily obvious that these features exist.

Basically I want to make these powerful tab-completion capabilities very discoverable for new ``cmd2`` users as I believe they are something which really set ``cmd2`` apart from competing alternatives.

**Comment by anselor** [comment]
@tleonhardt 

Instead of updating the example in README,md, I'm thinking I'll change how cmd2 handles all with_argparse commands such that if a complete_COMMAND function isn't provided cmd2 will automatically create an AutoCompleter. Users can still optionally implement the complete_CCOMMAND function to do whatever they want.

**Comment by tleonhardt** [comment]
@anselor 
I like your idea much better ;-)  

If you just make it so the two ``argparse``-based decorators automatically create an AutoCompleter that sounds like the best possible scenario!

**Comment by tleonhardt** [comment]
@anselor 
We currently don't have macOS CI testing enabled because it is a bit unreliable on TravisCI (sometimes it gets badly backed up).  I did a little manual testing on your branch on my Mac and got a bunch of failures when running with Python 3.6, but everything worked fine when running with Python 3.5.  

When I had problems, it was apparently entirely within the **test_acargparse.py** unit test file.

Errors tended to look like the following:
```Bash
========================================================= ERRORS =========================================================
_____________________________________ ERROR collecting tests/test_autocompletion.py ______________________________________
tests/test_autocompletion.py:20: in <module>
    from examples.tab_autocompletion import TabCompleteExample
examples/tab_autocompletion.py:13: in <module>
    class TabCompleteExample(cmd2.Cmd):
examples/tab_autocompletion.py:94: in TabCompleteExample
    suggest_parser = AutoCompleter.ACArgumentParser()
cmd2/AutoCompleter.py:717: in __init__
    super().__init__(**params)
E   TypeError: __init__() got an unexpected keyword argument 'allow_abbrev'
```

So I would strongly advise you borrow a Mac for some testing before you merge this in.

**Comment by anselor** [comment]
The allow_abbrev constructor parameter was added in python 3.5. I'm suspicious about what version of python is executing on your system. I think it would only be unrecognized if you were running 3.4 or 2.7.

That being said, I changed the way I pass through constructor parameters so it shouldn't show up again. This also reduces the maintenance overhead of having to match the parent constructor's parameters.

**Comment by tleonhardt** [comment]
I'm now getting a different error with the latest code on your branch, but again only for Python 3.6 not Python 3.5:
```Bash
========================================================= ERRORS =========================================================
_____________________________________ ERROR collecting tests/test_autocompletion.py ______________________________________
tests/test_autocompletion.py:20: in <module>
    from examples.tab_autocompletion import TabCompleteExample
examples/tab_autocompletion.py:13: in <module>
    class TabCompleteExample(cmd2.Cmd):
examples/tab_autocompletion.py:104: in TabCompleteExample
    def do_suggest(self, args) -> None:
cmd2/cmd2.py:309: in arg_decorator
    cmd_wrapper.__doc__ = argparser.format_help()
cmd2/AutoCompleter.py:751: in format_help
    formatter.add_arguments(req_args)
../../anaconda/lib/python3.6/site-packages/argparse.py:300: in add_arguments
    self.add_argument(action)
../../anaconda/lib/python3.6/site-packages/argparse.py:285: in add_argument
    invocations = [get_invocation(action)]
cmd2/AutoCompleter.py:634: in _format_action_invocation
    default = self._get_default_metavar_for_optional(action)
E   AttributeError: 'ACHelpFormatter' object has no attribute '_get_default_metavar_for_optional'
```

**Comment by anselor** [comment]
I'm suspicious about what version of python is running on your mac when you run the test. I checked and `_get_default_metavar_for_optional` is present on 3.4, 3.5, and 3.6. It isn't, however, on 2.7. That would also explain why it didn't recognize the `allow_abbrev` parameter in the constructor. Maybe try uninstalling 2.7 and seeing what happens?

**Comment by tleonhardt** [comment]
Its definitely 3.6.  Borrow the "Interview" MacBook Pro off of Kevin and test on there, it is setup nearly identically to mine.

**Comment by anselor** [comment]
The change to enable AutoCompleter by default for all commands that use argparse requires some reworking of how cmd2 handles argparse subcommands. I think I'd rather get the current set of changes in and split that work into a second round of changes. Sound good?

**Comment by tleonhardt** [comment]
Yeah, splitting it into two PRs sounds like a good idea.

**Comment by anselor** [comment]
@tleonhardt Looks like this is a side effect of turning cmd2 into a package. When running cmd2.py directly, it's not running in the context of a package so it can't perform package-local imports. When loading cmd2 externally, it's loading cmd2.py in the context of the package so package-local imports are required.

Two possible solutions I've found:
1 - Run cmd2 as a package:  `$ python3 -m cmd2.cmd2`
Generates this warning:
`/usr/lib/python3.5/runpy.py:125: RuntimeWarning: 'cmd2.cmd2' found in sys.modules after import of package 'cmd2', but prior to execution of 'cmd2.cmd2'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))`

2 - Move the code in `__main__` into `main.py` at the top level of the project that just imports and runs cmd2

@kotfu Maybe you have another suggestion for how to deal with this?

**Comment by tleonhardt** [comment]
@anselor 
I don't know what was going on with the unit test errors.  But I created a new clean Python 3.6.5 virtual environment with minimal requirements installed and the unit tests run fine.  So maybe something is messed up in my default environment.

Regarding now being able to run **cmd2.py** directly anymore, that is fine.  I don't need to be able to do that.  I can use any one of the examples for manual testing.

**Comment by anselor** [comment]
@tleonhardt I suspect there's some sort of pathing problem in your virtual environment. The behavior suggests it's loading the 2.7 version of argparse.py in the Python 3.6 interpreter.
